Yt integrations

Eliza-Diggins opened the pull request:
# Chunking For YT dataset generation


## Features

- Improvements to the YT dataset generation process.
  - Added ``create_dataset`` method to ``ClusterModel`` class.
  - Modified the ``create_dataset`` algorithm in ``ClusterICs`` class.
  - Rewrote ``data_structures`` to allow for two routes by which to generate ``yt`` datasets:
    - A: Dataset is produced as an HDF5 file containing the relevant grids;
      - Computations are carried out in chunks to reduce memory load and allow for larger overall grids.
      - User can select to either write HDF5 to disk (can then read the dataset in directly from ``yt``) or generate the HDF5 dataset in a temp directory to be deleted after the current script has been completed.
    - B: Dataset is created in memory as was done in the past. Obvious limitations on memory usage for the dataset.
    
## Bug Fixes

- Fixed miscellaneous issues with the IC level ``create_dataset`` procedure.
- Removed various functions in ``utils.py`` which were originally purposed for later use but are now unneeded. 

## Documentation

- API documentation is complete for these additions.
- Examples / more complete documentation will be added in a future pull request. [Later planned pull requests have better documentation setup than current branch].

Commit:
Added chunking to dataset initialization.

Commit:
Added chunking capabilities for large YT dataset generation.

jzuhone commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,310 @@
+            ] = rr
+
+        del x, y, z, rr
+        gc.collect()
```
@Eliza-Diggins are you sure we need to garbage-collect by hand here? It's very rare that one needs to call this explicitly in packages. It usually indicates that we should re-write things to avoid it.

Eliza-Diggins replied:
@jzuhone I don't think its strictly necessary.

jzuhone commented:
@Eliza-Diggins I took a first pass look over this--I think that this is definitely the right idea. However, I am wondering if we should opt for a somewhat simpler approach, which involves the following:

1. Always chunk. I can't think of an instance where I wouldn't want to just chunk by default. It will make the code logic simpler. 
2. Another way to straightforwardly simplify the logic and to avoid memory issues is to always save to an HDF5 file. This makes the solution portable. 
3. Finally, I think that we could also avoid memory issues and retain execution speed if we took the code that loops over the chunked grids and put it into Cython/C. Within Cython we could do the loop over grids, and the code that actually interpolates to the grid cells could be written entirely in C. We could literally just adapt this function from the GAMER code: https://github.com/gamer-project/gamer/blob/01b060a112b522f1407f62d897a1715b157ae580/src/TestProblem/Hydro/ClusterMerger/Init_TestProb_ClusterMerger.cpp#L479 The advantage of doing this is that you don't need to create large 3D arrays of coordinates--you just use the 1D arrays of coordinates as you iterate over the grids. 

I'm happy to help out with this last part, as I'm not sure if you've used Cython before.

Eliza-Diggins commented:
> @Eliza-Diggins I took a first pass look over this--I think that this is definitely the right idea. However, I am wondering if we should opt for a somewhat simpler approach, which involves the following:
> 
> 1. Always chunk. I can't think of an instance where I wouldn't want to just chunk by default. It will make the code logic simpler.
> 2. Another way to straightforwardly simplify the logic and to avoid memory issues is to always save to an HDF5 file. This makes the solution portable.
> 3. Finally, I think that we could also avoid memory issues and retain execution speed if we took the code that loops over the chunked grids and put it into Cython/C. Within Cython we could do the loop over grids, and the code that actually interpolates to the grid cells could be written entirely in C. We could literally just adapt this function from the GAMER code: https://github.com/gamer-project/gamer/blob/01b060a112b522f1407f62d897a1715b157ae580/src/TestProblem/Hydro/ClusterMerger/Init_TestProb_ClusterMerger.cpp#L479 The advantage of doing this is that you don't need to create large 3D arrays of coordinates--you just use the 1D arrays of coordinates as you iterate over the grids.
> 
> I'm happy to help out with this last part, as I'm not sure if you've used Cython before.

@jzuhone That's a good plan. I'm relatively comfortable about writing the Cython aspects of the code, but I may need to lean on you if I get stuck! I'll dig into it and see what I can do.

Commit:
working on the cython implementation.

Eliza-Diggins commented:
@jzuhone Hey John, A few questions for you:

- When writing to the grids in HDF5 files, its easy enough to have Cython access the *Python type* HDF5 buffer and write, but I requires repeated calls to the python interpreter. Other than that, everything can be done effectively in pure C. Is it worth trying to access the HDF5 C backend via h5py or should I just let the Cython call to the python object? 
- When ``YT`` is loading data from HDF5, I'm wondering how one assures that fields get read correctly. I.E. how to make sure YT interprets a grid ``dm_velocity_x`` as the field ``["dm"."velocity_x"]``?

jzuhone commented:
@Eliza-Diggins this was my (very rough) thought of how to do this:

1. Loop over fields in Python, allocate an array of shape `(ngrids, nx, ny, nz)`
2. For each field, pass this array to a Cython routine by reference (or pure C routine called by Cython) that iterates over the grids and fills up the field values. This routine simply defines coordinates as it goes (e.g., `x_i = x_0 + i*dx` or something like that) which obviates the need to allocate coordinate arrays. 
3. Array which has been filled gets written to HDF5 file by `h5py` back in Python. We then exit this iteration of the loop and do the next one. 

How's that sound?

jzuhone commented:
The other thing about the above is that it is easily extendable to AMR eventually

Commit:
working on the cython implementation.

Eliza-Diggins commented:
@jzuhone The newest commit is a considerable improvement and streamlines a lot of the logic. We how have a ``YTHDF5`` class in the ``data_structures`` module that handles all of the read / write aspects and eliminates most of the boilerplate that was needed in the previous version.

We now interpolate and write to h5 within the ``cluster_generator.opt.structures`` cython module.

The speeds are much improved; however, I think there are further optimizations to make. If you've got the time to play around with it, I'd love to see what a more practiced hand can squeeze out in terms of optimization!

jzuhone commented:
@Eliza-Diggins i started looking at this--I'm going to try to play around with it more next week.

jzuhone commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,560 @@
+    "velocity_x": ("density", 0, "velocity"),
+    "velocity_y": ("density", 1, "velocity"),
+    "velocity_z": ("density", 2, "velocity"),
+    "stellar_velocity_x": ("stellar_density", 0, "velocity"),
```
Should star and DM velocity fields be here? I actually don't think so. They're not going to be grid-based quantities. It may make sense to deposit the star and DM density fields, but not these.

jzuhone commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,560 @@
+        self._gridbuffer = None
+
+        if filename is None:
+            tf = tempfile.NamedTemporaryFile(delete=False)
```
Shouldn't we always require a file to be specified? Or is there a case where we would want a temp file?

jzuhone commented:
Hi @Eliza-Diggins: 

> The processes all work, but if you take a YTHDF5 object and create a dataset in yt, the HDF5 file gets stuck in r mode and you can't continue to edit the YTHDF5 object.

I think here you are stuck. I think you need to add everything to the dataset that you want and then at the end you can `load` it in yt.

> The fields all end up as (stream,*) once in yt. Obviously it would be nice to give yt a better idea of what the fields mean, but I'm not well enough versed in yt's backend to do so.

We can create a special frontend for `cluster_generator` datasets to handle this--in fact, we don't even have to put it in yt. We can make it a plugin in `cluster_generator` itself. But we can worry about that after this.

> I think the major bottleneck is the IO speed; which would make me think there should be a way to pass the read / write part of the code off to a multithreaded routine, but I haven't had the time / know-how to do so.

I wonder about a different method to handle this, which is inspired by how it's done for simulation codes. Consider the following idea (assuming that you are starting from scratch):

1. Open the HDF5 dataset for writing using h5py in Python.
2. In Python, start a loop over fields. For each field, write an array of zeros straight to disk (just write `np.zeros(...)` so it doesn't hang around in memory) corresponding to the field data, which is of the shape `(ngrids, nx, ny, nz)`. `ngrids` is the number of chunks, and `nx`, `ny`, and `nz` are the number of cells in each dimension for a single chunk/grid. I'm assuming that this is the same for every grid of course. You might have to call `flush` on the HDF5 dataset object after this. This step can be skipped if you're writing another cluster to an already existing dataset.
3. Loop over the chunks/grids in Python. For each grid, allocate an array of shape `(nx, ny, nz)` and pass it to a Cython routine that populates it with data from the profile(s). 
4. After this grid is filled up, access the part of the dataset you wrote earlier with shape `(ngrids, nx, ny, nz)` and add the part from this grid only, like:
```python
f["density"][igrid,:,:,:] = grid[:,:,:]
```
I am pretty sure that this does not require the whole HDF5 dataset to be read back into memory.

You'd loop over all of the fields in this way. If we end up having to parallelize, using `mpi4py` is probably the best option. Does the above sound feasible?

jzuhone commented:
@Eliza-Diggins one other thing I thought of that would simplify things tremendously--we should write only the minimal amount of data that we need to derive the rest, and we should only write fields that can be added, without needing to normalize. 

So the fields you would write are:

- Density
- Dark matter density
- Stellar density 
- Pressure
- Momentum density x,y,z (write density*velocity for all three)
- magnetic pressure (B^2/(4*pi)

That's all you need. These are all per-volume fields so if you want to add another cluster later you just need to add to the existing data and there is no need to normalize / un-normalize anything.

Everything else can be derived from those fields, and we can do it in the custom yt frontend I was talking about. Saves disk space and a lot of calculation.

Eliza-Diggins commented:
> @Eliza-Diggins one other thing I thought of that would simplify things tremendously--we should write only the minimal amount of data that we need to derive the rest, and we should only write fields that can be added, without needing to normalize.
> 
> So the fields you would write are:
> 
> * Density
> * Dark matter density
> * Stellar density
> * Pressure
> * Momentum density x,y,z (write density*velocity for all three)
> * magnetic pressure (B^2/(4*pi)
> 
> That's all you need. These are all per-volume fields so if you want to add another cluster later you just need to add to the existing data and there is no need to normalize / un-normalize anything.
> 
> Everything else can be derived from those fields, and we can do it in the custom yt frontend I was talking about. Saves disk space and a lot of calculation.

Hey John,

Sorry I've been slow getting around to this! I think implementing these changes should be simple enough; I'll hopefully have time next week!

Commit:
[! Not production ready]
Improved the data_structures.py logic for speedups.

Eliza-Diggins commented:
@jzuhone The latest commit should reflect those changes. It's still not blazing fast, but it's maxing out my laptop's IO capabilities. The new block structure of the HDF5 grids will (I think) require a custom YT frontend.

jzuhone commented:
Hi @Eliza-Diggins sorry (yet again sorry) for the delays--between XRISM proposals, meetings, and trying to figure out what to do about my own job (Chandra) situation I've fallen behind. But I've got some time opened up and I want to look at this during this week. Thanks for all of the hard work.

jzuhone commented:
@Eliza-Diggins ok I left some comments, nothing major. I will try this in the next few days on my own.

Eliza-Diggins commented:
@jzuhone No worries! I was sorry to hear the news about Chandra!

Let me know how things run for you. I'm hoping I'll have some time to work more on this and other aspects of the code in the coming few months!

jzuhone commented:
@Eliza-Diggins I started playing around with this but got some Cython errors--did it work for you?

Eliza-Diggins commented:
@jzuhone I believe it was running when I was last working on it; I'll dig into it and figure out what's going on. Was the error on installation / compilation or during execution?

Eliza-Diggins commented:
> @jzuhone I believe it was running when I was last working on it; I'll dig into it and figure out what's going on. Was the error on installation / compilation or during execution?

@jzuhone Ah! It definitely looks like I made a mistake somewhere! I'll patch the bug and add an implementation to the testing suite to make sure!

Commit:
Fixed various bugs with yt-integration, wrote unit test for data_structures.py, added documentation and improved file structure.

Eliza-Diggins commented:
@jzuhone Okay! Turns out I made several mistakes when I first wrote this, but things appear to be working nicely now. The new ``/tests/.test_data_structures.py`` test suite has a very basic unit test on the construction of the dataset.

At this point, the following are completed:

- Critical fields of ``ClusterModel`` classes are written to ``hdf5`` in chunks to avoid memory overload issues.
- Users can add new models to already existing ``YTHDF5`` objects with ease and ``ClusterICs`` can be added as well.
- Runtime ~ 5 min on a single core of my crappy laptop; I'm going to consider that a victory.

What still needs to happen:
- The arrays are written to ``hdf5`` in arrays which utilize the chunk structure (so ``yt`` can't naively read them). I think this will mean we need a custom ``yt`` frontend right? I'll look into doing this when I get a chance!
- AMR seems like an obvious improvement here, but complex to implement. Is there a standard structure for writing HDF5 files with AMR grids? Presumably the whole chunking aspect of things becomes nightmare-ish with multiple grid levels?

jzuhone commented:
Hi @Eliza-Diggins, 

Ok thanks!

As far as AMR goes, in general the idea is that a dataset array in the HDF5 file has a shape of something like `[ngrids,nx,ny,nz]`, where `nx,ny,nz` are the same for every grid. This easily generalizes to different refinement levels since you just need to store the metadata about the grid widths and locations as well.

I'll have another look at this soon.

jzuhone commented:
@Eliza-Diggins mind if I push a commit to this?

Eliza-Diggins commented:
> @Eliza-Diggins mind if I push a commit to this?

@jzuhone Not at all!

jzuhone committed:
Minor bugfixes. Also udpate pre-commit-config

Eliza-Diggins commented:
@jzuhone It looks like the next logical step for this PR is to implement a yt frontend to load the implemented data format. I'm wondering what the best approach would be for this? 

``yt`` has frontend guidelines / skeletons [here](https://yt-project.org/doc/developing/creating_frontend.html), which I think I could probably follow (albeit slowly), but its unclear to me if we would be better off with an extension written independently or if the frontend should be integrated into ``cluster_generator`` directly.

jzuhone commented:
@Eliza-Diggins i think we should put the frontend in `cluster_generator`. I tried this on my own last week and seemed to work well.

I can help with the frontend generation if you want. It should be a lot like FLASH or Athena++.

Commit:
[! Not production ready]
First pass at the frontend. The Dataset class is functional, but the io still needs to be written.

jzuhone commented:
@Eliza-Diggins I left some comments. Why not just put it under `cluster_generator/frontend`?

Commit:
added field aliases

Eliza-Diggins committed:
Delete cluster_generator/_extensions/cluster_generator_frontend directory

Commit:
added field aliases

Eliza-Diggins commented:
@jzuhone In a shocking twist, this actually appears to work! I've added the relevant aliases I could think of to ``fields.py`` and I'm able to get images out.

I'm not entirely confident about all of the ``yt``-side code (lots of guesswork and copy/paste) so it would probably be prudent to make sure I haven't done something entirely foolish!

Commit:
fixing github workflow issues.

Commit:
requiring h5py in setup.py

Commit:
fixed cumtrapz renamed in scipy 1.14.0

Commit:
fixed numpy complaint about not importing array in cython.

Commit:
trying to fix change in dtype size (likely due to numpy 2.0)

Commit:
removing --gravity (not present in this version of the code).

Commit:
working on improving the testing system

Commit:
Made significant changes to testing and utils.py

Commit:
Fixed 3.11 - <3.11 breaking on import of Self.

Commit:
Fixed issue with https://github.com/numpy/numpy/pull/10615.

Commit:
working on testing debugging.

Commit:
working on testing debugging.

Commit:
working on testing debugging.

Commit:
working on testing debugging.

Commit:
working on testing debugging.

Commit:
last checks on create_dataset functions. Added documentation.

Commit:
Fixing circular import

Commit:
fixing type hinting issue.

Commit:
Update README.rst

fixed the yt_project badge.

Eliza-Diggins committed:
Update README.rst

fixing some stupid bug with underscores. See https://github.com/scikit-learn/scikit-learn/pull/28815...

jzuhone commented:
Hi @Eliza-Diggins -- my example seems to work now. But did the frontend disappear? I seem to be missing it. 

Also, I had some unresolved questions above.

Eliza-Diggins commented:
Hey @jzuhone, the frontend should be tucked away in ``/cluster_generator/frontend`` now. 

I saw you mentioned leaving some comments, but I can't seem to see them on my end? I suspect user error on my part!

jzuhone commented:
@Eliza-Diggins sorry, weird GitHub interface. Check now

Eliza-Diggins commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,364 @@
+        )
+        _relative_bbox = self.bbox - center.d.reshape((3, 1))
+
+        mylog.info(f"Adding {model} to {self.__str__()}")
```
I didn't think about the performance issues that come from pre-computed f-strings in skipped logging statements. Will check around for any other examples of this that need fixing!

Eliza-Diggins commented on cluster_generator/_extensions/cluster_generator_frontend/data_structures.py:
```diff
@@ -0,0 +1,203 @@
+        self.length_unit = self.quan(1.0, "kpc")
+        self.mass_unit = self.quan(1.0, "Msun")
+        self.time_unit = self.quan(1.0, "Myr")
+        self.velocity_unit = self.quan(1.0, "km/s")
```
It is! Good catch!

Eliza-Diggins commented on cluster_generator/_extensions/cluster_generator_frontend/data_structures.py:
```diff
@@ -0,0 +1,203 @@
+    def __init__(
+        self,
+        filename,
+        dataset_type="cluster_generator_hdf5",
```
fixed in later commit.

Eliza-Diggins commented on cluster_generator/_extensions/cluster_generator_frontend/data_structures.py:
```diff
@@ -0,0 +1,203 @@
+        # fluid type or particle type.  Convention suggests that the on-disk
+        # fluid type is usually the dataset_type and the on-disk particle type
+        # (for a single population of particles) is "io".
+        self.field_list = [
```
Good catch! Fixed.

Eliza-Diggins commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,364 @@
+
+
+if __name__ == "__main__":
+    with YTHDF5.temporary() as temp:
```
:laughing: I'm not sure I understand the utility of it anymore. I think when I wrote it I was thinking about keeping the working directory from getting cluttered if you needed many of them, but this comes with other issues.

If the user makes an error inside the ``with`` statement, it could live in their temp space forever which is also not ideal. I think it makes more sense to let people implement this logic for themselves... 

I've removed it. Easy enough to add back if it ever makes sense to do so...

Eliza-Diggins commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,364 @@
+
+    def _add_field(self, fileio, r, y, fieldname, bbox, chunkmap):
+        # Interpolate and pass to the Cython level.
+        _, _, _, _, _, k, _, n, t, c, _, _, _, _ = dfitpack.fpcurf0(
```
I concur. It seems like there's a lot of active development in ``scipy.interpolation`` at the moment, so it's likely to need doing at some point...

Eliza-Diggins commented on cluster_generator/data_structures.py:
```diff
@@ -1,100 +1,364 @@
+        ), "The chunksize does not evenly divide the domain."
+
+        # Managing file existence logic.
+        if (filename.exists()) and (not overwrite):
```
Nope, fixed!

Commit:
various minor changes and style fixes.

Commit:
Trying to cherry pick

Eliza-Diggins commented:
> @Eliza-Diggins sorry, weird GitHub interface. Check now

Okay @jzuhone, I think I caught all of these and got them fixed up!

jzuhone commented:
@Eliza-Diggins I tried this out (and though I had to make one manual fix) it looks like everything is working! This is great. 

It looks like there are some merge conflicts. Can you fix those--and then I can push up the fix I have?

Eliza-Diggins commented:
@jzuhone , I think I managed to get those to merge correctly... I always think I'm good with git until I realize I'm not! Anyways, I think everything should be ready to go.

jzuhone commented:
Hi @Eliza-Diggins -- I saw that there were changes to a number of other things unrelated to the dataset generation. I hadn't appreciated how extensive they were until late last week (looking particularly in https://github.com/jzuhone/cluster_generator/pull/4/commits/3e5ef8109264b9ab20b595144fcd60c5312fd6ff, and in all of the changes to testing). 

I'm sorry I hadn't noticed all of it until late, but this is too much for one PR, and it's unrelated to the main purpose of the PR in any case. Is it possible for you to rebase, preserving those commits somehow in a separate branch (by cherry-picking, probably?) I don't want you to lose the work, but major changes should really have their own proper and separate review as well as a broader discussion about code design. 

Not questioning the value of any of it at all--just want to keep things more manageable for reviewing.

Eliza-Diggins commented:
Hey @jzuhone , no worries, I'll rebase as best I can. I think it may be tricky to reverse some of the documentation changes, but they were largely just adding documentation where there wasn't any. The utilities module and the testing module can certainly be dealt with separately. I'll make pull requests for those.

Commit:
Trying to cherry pick

Commit:
Trying to cherry pick

Commit:
Reverting changes to various files.

Commit:
Fixing some testing issues to get minimal passing repo.

Commit:
fixed minor doctest issue. Tests now passing locally.

Eliza-Diggins commented:
@jzuhone Okay, that should be much, much less messy to review. There are some spots where I added documentation; I've left those, but they aren't all that frequent. Most of the other changes were necessary to get tests passing on github actions (dealing with version inconsistencies, etc.). All of the massive revamping of the ``utils`` and ``tests`` are gone with the exception of the testing changes needed for ``test_data_structures.py``. Let me know if you need anything else on this one!

jzuhone committed:
Fixing docstrings and some stuff that got missed in a merge

jzuhone committed:
Bugfix

jzuhone committed:
We don't need this to show up unless we're debugging

jzuhone commented:
@Eliza-Diggins sorry this took me so long, pushed up some fixes and merging now. great stuff!
